src/main/java/backend/Logic.java

```diff
@@ -196,18 +191,36 @@ public Model getRepo(String repoId) {
      * @return True if label replacement on GitHub was a success, false otherwise.
      */
     public CompletableFuture<Boolean> replaceIssueLabels(TurboIssue issue, List<String> newLabels) {
+        eagerlyUpdateIssueLabels(issue, newLabels);
+
         logger.info("Changing labels for " + issue + " on GitHub");
-        return repoIO.replaceIssueLabels(issue, newLabels)
-                .thenApply(labels -> {
-                    logger.info("Changing labels for " + issue + " on UI");
-                    issue.setLabels(labels);
-                    refreshUI();
-                    return true;
-                })
+        return repoOpControl.replaceIssueLabels(issue, newLabels)
+                .thenApply(labels -> labels.containsAll(newLabels))
                 .exceptionally(Futures.withResult(false));
     }
 
     /**
+     * Assigns a list of labels to an issue offline and immediately reflect the change on the UI without
+     * concerning whether the labels have been updated on the remote repo's data source
+     *
+     * The UI will not be force refreshed if the set of labels are the same as the existing labels
+     * assigned to the issue.
```

Comment:
The `isLabelsUnchanged` check is mostly used to avoid doing `refreshUI` so they need to tag along together. One way is to make the Ui "observe" the issue somehow, then when we call `issue.setLabel` etc. it knows to update, but that's kind of far off. Another way is just to bring the whole block back up i.e. to `replaceIssueLabels`. What do you think?